pallets/offchain-task-scheduler/src/authority.rs

```diff
@@ -0,0 +1,49 @@
+use super::pallet;
+use super::Config;
+
+pub trait AuthorityController<AccountId> {
```

Comment:
IMO `AccountId` makes more sense as an associated type rather than a trait parameter, since each runtime has only one sensible `AccountId` type (`Runtime::AccountId`).

Also, if you ever need to be generic over `AuthorityController`s, the bounds are nicer with associated types:

```rust
fn uses_authority_controller<C: AuthorityController>() {}
```
vs
```rust
fn uses_authority_controller<F, C: AuthorityController<F>>() {}
```


(Similarly, if you ever need to fully-qualify the type it's nicer as well: `<Runtime as AuthorityController<<Runtime as Config>::AccountId>>::is_authority(...))` vs `<Runtime as AuthorityController>::is_authority(...)`).

Reply:
@nathanwhit Good catch, I agree but, correct me if I'm wrong.
The thing is, I really like `AuthorityController` being an [auto trait](https://doc.rust-lang.org/beta/unstable-book/language-features/auto-traits.html#items), this way we don't have to implement the trait in the runtime for every runtime, not that we are going to have more than a few ever.

I don't see the trait bound or the fully qualified syntax as a problem because everywhere you use the trait, you are generic over the Runtime anyway and with the Config's associated types the compiler infers the generic call. I don't expect having to ever use the fully-qualified syntax for more or less the same reason, unless we have to disambiguate the call, but what are the odds?
Look for `AuthorityController` in #773 and how you never see it outside of its definition crate besides being a Config associated type bound and being imported. Let me know if this is not enough to convince you.

Reply:
Hmm it's not quite an auto trait, it's just a blanket impl. (User-defined auto traits are still unstable, and AFAIK will probably never be stabilized. The difference is that auto traits like `Send`, `Sync`, and `Sized` don't have an explicit impl, an impl is automatically generated by the compiler depending on the properties of the type). 

You can still write an equivalent blanket impl even when `AccountId` is an associated type:

```rust
pub trait AuthorityController {
    type AccountId;
	fn insert_authority(authority: &Self::AccountId);
	fn remove_authority(authority: &Self::AccountId);
	fn is_authority(authority: &Self::AccountId) -> bool;
}

impl<Runtime: Config> AuthorityController for Runtime {
    type AccountId = Runtime::AccountId;
	fn insert_authority(authority: &Runtime::AccountId) {
		pallet::Authorities::<Runtime>::insert(authority, ());
	}
	fn remove_authority(authority: &Runtime::AccountId) {
		pallet::Authorities::<Runtime>::remove(authority);
	}
	fn is_authority(authority: &Runtime::AccountId) -> bool {
		pallet::Authorities::<Runtime>::contains_key(authority)
	}
}
```

Reply:
Right blanket traits, I always confuse the two. Silly me, of course it's doable lol.